### PR TITLE
[3.7] Add restart single master handler for etcdv3 upgrade from embedded etcd

### PR DESCRIPTION
**Pull request by nak3**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1564098

### CODE REVIEW

**Comment by vrutkovs** [comment]
Not sure why would we want to support embedded etcd - the first task in migrate is verifying if embedded etcd is used - and requires to migrate to external etcd, see https://github.com/openshift/openshift-ansible/blob/release-3.7/playbooks/common/openshift-etcd/migrate.yml#L11

**Comment by nak3** [comment]
The customer was supposed to finish the migration to external etcd, you can refer to following logs attached in https://bugzilla.redhat.com/show_bug.cgi?id=1564098.

```
PLAY [Check if the master has embedded etcd] ******************************************************************************************************************************************************************

TASK [fail] ***************************************************************************************************************************************************************************************************
skipping: [localhost]

PLAY [Run pre-checks] *****************************************************************************************************************************************************************************************
```

But their master is still single. Do you mean that embedded2external.yml should have doen the change of separation from master to `master-api`/`mater-controllers`?

**Comment by vrutkovs** [comment]
Oh, correct, my fault. I was under impression that `api/controller` is now enforced in 3.7.

/lgtm

**Comment by nak3** [comment]
`openshift-jenkins/system-containers` has an issue. I opened https://github.com/openshift/openshift-ansible/issues/7832

**Comment by nak3** [comment]
That makes sense. Thank you. Updated. (Don't know why logging test started failing...)

**Comment by nak3** [comment]
ci/openshift-jenkins/logging also has an issue in 3.7 branch. These PR have same issue with this PR.

https://github.com/openshift/openshift-ansible/pull/7875
https://github.com/openshift/openshift-ansible/pull/7891
https://github.com/openshift/openshift-ansible/pull/7892

**Comment by DanyC97** [comment]
@nak3 so what is the issue with #7875  & ci/openshift-jenkins/logging ?

**Comment by nak3** [comment]
@DanyC97 It is same issue with https://github.com/openshift/origin/issues/18184
Upstream (master branch) fixed the issue, but release-3.7 didn't fix it. I requested if it is possible to merge it into release-3.7 as https://github.com/openshift/origin/pull/18647#issuecomment-380465245

File: roles/openshift_master/handlers/main.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@
   when:
   - not (master_api_service_status_changed | default(false) | bool)
   - openshift.master.cluster_method == 'native'
+  - master_api_enabled.rc == 0 or openshift_version | version_compare('3.7','>=')
```
Comment (sdodson) [inline roles/openshift_master/handlers/main.yml]: I wonder if there may be cases where this code is called and this variable may be undefined? Perhaps
` - master_api_enabled is defined and (master_api_enabled.rc == 0 or openshift_version | version_compare('3.7','>='))`

File: roles/openshift_master/handlers/main.yml [hunk 2]
```diff
@@ -19,6 +20,18 @@
...
+  - master_controllers_enabled.rc == 0 or openshift_version | version_compare('3.7','>=')
+
+# For now, this handler is called only from update_etcd_client_urls.yml in case migration from previous version.
+# Thus, {{ *_service_status_changed }} like api, controllers is neither implemented nor handled.
+- name: restart single master
+  systemd:
+    name: "{{ openshift.common.service_type }}-master"
+    state: restarted
+  when:
+  - single_master_enabled.rc == 0 or openshift_version | version_compare('3.7','<')
```
Comment (sdodson) [inline roles/openshift_master/handlers/main.yml]: likewise, should we check for undefined here?

File: roles/openshift_master/handlers/main.yml (hunk 1 above)
Comment (nak3) [reply roles/openshift_master/handlers/main.yml]: Thank you. Yes, it is safer.

I wondered if I should write these as:

~~~
  - master_api_enabled is defined
  - master_api_enabled.rc == 0 or openshift_version | version_compare('3.7','>=')
~~~

But maybe `- master_api_enabled is defined and (master_api_enabled.rc == 0 or openshift_version | version_compare('3.7','>='))` is easier to understand why `master_api_enabled is defined` is checked.

**Comment by vrutkovs** [comment]
Verified that it works without `openshift_master_cluster_method` set and with `openshift_master_cluster_method: native`

**Pull request merged**